File: src/commands/tasks.ts
/**
 * mm tasks — read + write Google Tasks across all the user's lists.
 *
 *   mm tasks                           due in next 7 days, all lists (= `list`)
 *   mm tasks list [--days N] [--all]
 *   mm tasks add "Buy milk" [--due "tomorrow"] [--list "Home"] [--notes "..."]
 *   mm tasks done <task-id> [--list-id <list-id>]
 *
 * `--due` accepts natural language ("tomorrow", "next friday") or
 * ISO `YYYY-MM-DD`. Google Tasks stores due as a date only.
 */
import { hubApi } from '../hub';
import { parseNlDate } from '../nl-date';

type Task = {
	id: string;
	title: string;
	notes?: string;
	due?: string;
	status: 'needsAction' | 'completed';
	webViewLink?: string;
};

type Group = { listId: string; listTitle: string; tasks: Task[] };
type ListResp = { groups: Group[]; accountSlug: string | null };
type AddResp = { id: string; listId: string; listTitle: string };

export async function tasksDispatch(
	command: string,
	args: string[],
	flags: { json?: boolean }
) {
	const json = flags?.json || false;
	if (command.startsWith('--')) {
		args = [command, ...args];
		command = 'list';
	}
	switch (command) {
		case '':
		case 'list':
		case 'ls':
			return tasksList(args, json);
		case 'add':
		case 'new':
			return tasksAdd(args, json);
		case 'done':
		case 'complete':
			return tasksDone(args, json);
		default:
			console.error(`Unknown command: mm tasks ${command}`);
			process.exit(1);
	}
}

function parseFlags(args: string[]): { flags: Record<string, string>; positional: string[] } {
	const flags: Record<string, string> = {};
	const positional: string[] = [];
	for (let i = 0; i < args.length; i++) {
		const a = args[i];
		const eq = a.match(/^--([a-z][a-z-]*)=(.+)$/s);
		if (eq) {
			flags[eq[1]] = eq[2];
			continue;
		}
		const longFlag = a.match(/^--([a-z][a-z-]*)$/);
		if (longFlag) {
			// Boolean if next arg is also a flag, else value.
			const next = args[i + 1];
			if (next === undefined || next.startsWith('--')) {
				flags[longFlag[1]] = 'true';
			} else {
				flags[longFlag[1]] = next;
				i++;
			}
			continue;
		}
		positional.push(a);
	}
	return { flags, positional };
}

function fmtDue(iso: string | undefined): string {
	if (!iso) return '';
	const d = new Date(iso);
	if (Number.isNaN(d.getTime())) return iso;
	return d.toLocaleDateString('en-GB', { day: 'numeric', month: 'short' });
}

async function tasksList(args: string[], json: boolean) {
	const { flags } = parseFlags(args);
	const payload: Record<string, unknown> = {};
	if (flags.days) payload.days = Number(flags.days);
	if (flags.all) payload.all = true;
	if (flags.account) payload.accountSlug = flags.account;

	const data = await hubApi<ListResp>('tasks', 'list', payload);
	if (json) {
		console.log(JSON.stringify(data, null, 2));
		return;
	}

	const total = data.groups.reduce((n, g) => n + g.tasks.length, 0);
	if (total === 0) {
		console.log('Nothing pending. ✨');
		return;
	}

	for (const g of data.groups) {
		if (g.tasks.length === 0) continue;
		console.log(`\n${g.listTitle}  (${g.tasks.length})`);
		for (const t of g.tasks) {
			const due = t.due ? `  due ${fmtDue(t.due)}` : '';
			console.log(`  · ${t.title}${due}`);
			console.log(`      id=${t.id}  list=${g.listId}`);
			if (t.notes) console.log(`      ${t.notes.split('\n').join('\n      ')}`);
		}
	}
}

async function tasksAdd(args: string[], json: boolean) {
	const { flags, positional } = parseFlags(args);
	const title = positional.join(' ').trim();
	if (!title) {
		console.error('Usage: mm tasks add "<title>" [--due YYYY-MM-DD] [--list "List name"] [--notes "..."]');
		process.exit(1);
	}
	const payload: Record<string, unknown> = { title };
	let dueDisplay: string | undefined;
	if (flags.due) {
		try {
			const parsed = parseNlDate(flags.due);
			payload.due = parsed.iso;
			dueDisplay = parsed.iso;
		} catch (err) {
			console.error(`✗ ${err instanceof Error ? err.message : err}`);
			process.exit(1);
		}
	}
	if (flags.list) payload.listTitle = flags.list;
	if (flags['list-id']) payload.listId = flags['list-id'];
	if (flags.notes) payload.notes = flags.notes;
	if (flags.account) payload.accountSlug = flags.account;

	const data = await hubApi<AddResp>('tasks', 'add', payload);
	if (json) {
		console.log(JSON.stringify(data, null, 2));
		return;
	}
	console.log(`✓ Added to ${data.listTitle}`);
	console.log(`  ${title}`);
	if (dueDisplay) console.log(`  due ${dueDisplay}`);
}

async function tasksDone(args: string[], json: boolean) {
	const { flags, positional } = parseFlags(args);
	const taskId = positional[0];
	if (!taskId) {
		console.error('Usage: mm tasks done <task-id> [--list-id <list-id>]');
		console.error('       (run `mm tasks` first to see ids)');
		process.exit(1);
	}
	const listId = flags['list-id'];
	if (!listId) {
		console.error('Missing --list-id. `mm tasks` shows ids for each task — the line below the title.');
		process.exit(1);
	}

	const data = await hubApi<{ ok: true }>('tasks', 'complete', {
		listId,
		taskId,
		...(flags.account ? { accountSlug: flags.account } : {})
	});
	if (json) {
		console.log(JSON.stringify(data, null, 2));
		return;
	}
	console.log('✓ Done.');
}

File: internal/cmd/tasks.go
package cmd

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"mm-cli/internal/http"
	"mm-cli/internal/nldate"
	"mm-cli/internal/wire"
)

// NewTasksCmd builds the `mm tasks` tree.
func NewTasksCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "tasks",
		Short: "Google Tasks — list, add, complete",
		Args:  cobra.NoArgs,
		RunE:  runTasksList,
	}
	cmd.Flags().Int("days", 7, "Window in days")
	cmd.Flags().Bool("all", false, "Include all lists")
	cmd.Flags().String("account", "", "Pick a linked Google account")
	cmd.AddCommand(newTasksListCmd(), newTasksAddCmd(), newTasksDoneCmd())
	return cmd
}

func newTasksListCmd() *cobra.Command {
	c := &cobra.Command{Use: "list", Aliases: []string{"ls"}, Short: "List tasks", Args: cobra.NoArgs, RunE: runTasksList}
	c.Flags().Int("days", 7, "Window in days")
	c.Flags().Bool("all", false, "Include all lists")
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func newTasksAddCmd() *cobra.Command {
	c := &cobra.Command{
		Use:     "add [title]",
		Aliases: []string{"new"},
		Short:   "Add a task",
		Args:    cobra.MinimumNArgs(1),
		RunE:    runTasksAdd,
	}
	c.Flags().String("due", "", "Due date (NL or YYYY-MM-DD)")
	c.Flags().String("list", "", "Task list title")
	c.Flags().String("list-id", "", "Task list id")
	c.Flags().String("notes", "", "Long description")
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func newTasksDoneCmd() *cobra.Command {
	c := &cobra.Command{
		Use:     "done [task-id]",
		Aliases: []string{"complete"},
		Short:   "Mark a task complete",
		Args:    cobra.ExactArgs(1),
		RunE:    runTasksDone,
	}
	c.Flags().String("list-id", "", "Task list id (required)")
	c.Flags().String("account", "", "Pick a linked Google account")
	return c
}

func runTasksList(cmd *cobra.Command, _ []string) error {
	days, _ := cmd.Flags().GetInt("days")
	all, _ := cmd.Flags().GetBool("all")
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")

	req := map[string]any{}
	if days != 7 && days > 0 {
		req["days"] = days
	}
	if all {
		req["all"] = true
	}
	if account != "" {
		req["accountSlug"] = account
	}

	client := http.New()
	var resp wire.HubTasksListResp
	if err := client.Hub(cmd.Context(), "tasks", "list", req, &resp); err != nil {
		return err
	}

	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}

	total := 0
	for _, g := range resp.Groups {
		total += len(g.Tasks)
	}
	if total == 0 {
		fmt.Println("Nothing pending. ✨")
		return nil
	}
	for _, g := range resp.Groups {
		if len(g.Tasks) == 0 {
			continue
		}
		fmt.Printf("\n### %s (%d)\n", g.ListTitle, len(g.Tasks))
		for _, t := range g.Tasks {
			due := ""
			if t.Due != "" {
				due = " — *due " + fmtDue(t.Due) + "*"
			}
			fmt.Printf("- **%s**%s\n", t.Title, due)
			fmt.Printf("  - ID: `%s` | List: `%s`\n", t.ID, g.ListID)
			if t.Notes != "" {
				lines := strings.ReplaceAll(t.Notes, "\n", "\n    ")
				fmt.Printf("    > %s\n", lines)
			}
		}
	}
	return nil
}

func runTasksAdd(cmd *cobra.Command, args []string) error {
	title := strings.TrimSpace(strings.Join(args, " "))
	if title == "" {
		return fmt.Errorf("title is required")
	}
	due, _ := cmd.Flags().GetString("due")
	listFlag, _ := cmd.Flags().GetString("list")
	listID, _ := cmd.Flags().GetString("list-id")
	notes, _ := cmd.Flags().GetString("notes")
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")

	req := map[string]any{"title": title}
	var dueDisplay string
	if due != "" {
		parsed, err := nldate.ParseDate(due, time.Now())
		if err != nil {
			return err
		}
		req["due"] = parsed.ISO
		dueDisplay = parsed.ISO
	}
	if listFlag != "" {
		req["listTitle"] = listFlag
	}
	if listID != "" {
		req["listId"] = listID
	}
	if notes != "" {
		req["notes"] = notes
	}
	if account != "" {
		req["accountSlug"] = account
	}

	client := http.New()
	var resp wire.HubTasksAddResp
	if err := client.Hub(cmd.Context(), "tasks", "add", req, &resp); err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	fmt.Printf("✓ Added to %s\n  %s\n", resp.ListTitle, title)
	if dueDisplay != "" {
		fmt.Printf("  due %s\n", dueDisplay)
	}
	return nil
}

func runTasksDone(cmd *cobra.Command, args []string) error {
	taskID := args[0]
	listID, _ := cmd.Flags().GetString("list-id")
	account, _ := cmd.Flags().GetString("account")
	wantJSON, _ := cmd.Root().PersistentFlags().GetBool("json")
	if listID == "" {
		return fmt.Errorf("Missing --list-id. `mm tasks` shows ids for each task — the line below the title.")
	}
	req := map[string]any{"listId": listID, "taskId": taskID}
	if account != "" {
		req["accountSlug"] = account
	}
	client := http.New()
	var resp wire.HubTasksCompleteResp
	if err := client.Hub(cmd.Context(), "tasks", "complete", req, &resp); err != nil {
		return err
	}
	if wantJSON {
		out, _ := json.MarshalIndent(resp, "", "  ")
		fmt.Println(string(out))
		return nil
	}
	fmt.Println("✓ Done.")
	return nil
}

func fmtDue(iso string) string {
	if iso == "" {
		return ""
	}
	t, err := time.Parse("2006-01-02", iso)
	if err != nil {
		t, err = time.Parse(time.RFC3339, iso)
		if err != nil {
			return iso
		}
	}
	return t.Format("2 Jan")
}
